custom/opm/reports.py

```diff
@@ -1041,28 +1014,27 @@ def headers(self):
             headers.append(DataTablesColumn(name=title, help_text=text))
         return DataTablesHeader(*headers)
 
-    def get_matching_awcs(self):
-        filter_on = 'awc' if self.awcs else 'gp' if self.gp else 'block'
-        for block, gps in get_hierarchy().items():
-            if not filter_on == 'block' or block == self.block:
-                for gp, awcs in gps.items():
-                    if not filter_on == 'gp' or gp in self.gp:
-                        for awc in awcs.keys():
-                            if not filter_on == 'awc' or awc in self.awcs:
-                                yield awc
-
+    @property
+    @memoized
     def awc_data(self):
         case_objects = self.row_objects + self.extra_row_objects
-        awcs = {awc: [] for awc in self.get_matching_awcs()}
-        # populate awcs with cases from that awc
+        cases_by_owner = {}
         for case_object in case_objects:
-            awc = case_object.awc_name
-            awcs[awc] = awcs.get(awc, []) + [case_object]
-        return awcs
+            owner_id = case_object.owner_id
+            cases_by_owner[owner_id] = cases_by_owner.get(owner_id, []) + [case_object]
+        return cases_by_owner
+
+    def iter_awcs(self):
+        for user in self.users_matching_filter:
+            yield AWCHealthStatus(
+                cases=self.awc_data.get(user['doc_id'], []),
+                awc=user['awc'],
+                awc_code=user['awc_code'],
+                gp=user['gp'],
+            )
```

Comment:
I decided to pull this part out to be sure the row instantiation happens only in one place.